lib/puppet/type/exec.rb

```diff
@@ -204,7 +204,7 @@ def value=(*values)
       validate do |user|
         if Puppet.features.microsoft_windows?
           self.fail "Unable to execute commands as other users on Windows"
-        elsif !Puppet.features.root? && resource.current_username() != user
+        elsif Facter.value(:id) != 'root' && Facter.value(:id) != user
```

Comment:
@mcanevet so `it { should compile }` actually does two things, it compiles the "resource" catalog (which normally happens on the master) and it converts it to a "ral" catalog (which normally happens on the agent). In your case, both steps are happening together, and the same is true when doing `puppet apply`.

The second step is what triggers the validate method in the `user` parameter of the `exec` type. Since you are not running the test as root, the test fails as expected.

I believe the fix should be done in your test, e.g.

<pre>
before :each do
  Puppet.features.stubs(:root?).returns(true)
end

it { should compile }
</pre>